icu4c/source/common/unicode/unistr.h

```diff
@@ -3258,10 +3263,13 @@ class U_COMMON_API UnicodeString : public Replaceable
    * @param text UTF-16 string
    * @draft ICU 76
    */
-  template<typename S, typename = std::enable_if_t<ConvertibleToU16StringView<S>>>
-  explicit inline UnicodeString(const S &text) {
+  template <typename S,
+            typename = std::enable_if_t<
+                ConvertibleToU16StringView<S> &&
+                !std::is_same_v<std::remove_cv_t<std::remove_reference_t<S>>, UnicodeString>>>
+  UNISTR_FROM_STRING_EXPLICIT UnicodeString(S&& text) {
```

Comment:
Declaring the parameter as `S&&` goes together with then passing it on using `std::forward<S>()` and this is what's called _perfect forwarding_ and this is what delays the possible construction of any new string view object until after it has been verified that the parameter isn't a `nullptr`.

Reply:
> this is what delays the possible construction of any new string view object until after it has been verified that the parameter isn't a nullptr.

What? If S is not a string view already, but is convertible to a string view, then const S& is a const reference to an S, not a string view; the construction only happens when you pass it to a function that actually takes a string_view (so now, it happens when you call toU16StringView in toU16StringViewNullable).

Perfect forwarding is about preserving the value category, for instance, to allow a move constructor to be called through another function (as in the `emplace` member function on containers). I don’t think it helps here.

Reply:
> […]; the construction only happens when you pass it to a function that actually takes a string_view (so now, it happens when you call toU16StringView in toU16StringViewNullable).

Exactly, that was what I was trying to express.

Reply:
But you don’t need perfect forwarding to achieve that, a const S& here and a const T& in toU16StringViewNullable  should also work.

Reply:
I see that you also backed out the condition `!std::is_same_v<std::remove_cv_t<std::remove_reference_t<S>>, UnicodeString>`. Did you originally add that to yield to the copy constructor? Did you now conclude that you don't need it because that is a more direct signature match anyway?

Reply:
That condition was to not match the user-defined conversion function from `UnicodeString` to `std::u16string_view` which otherwise would have matched `ConvertibleToU16StringView`.